src/game.cpp

```diff
@@ -1533,16 +1533,32 @@ bool game::do_turn()
 
             mon_info_update();
 
-            // If player is performing a task and a monster is dangerously close, warn them
-            // regardless of previous safemode warnings
+            // If player is performing a task, a monster is dangerously close,
+            // and monster can reach to the player or it has some sort of a ranged attack,
+            // warn them regardless of previous safemode warnings
             if( u.activity && !u.has_activity( activity_id( "ACT_AIM" ) ) &&
                 u.activity.moves_left > 0 &&
                 !u.activity.is_distraction_ignored( distraction_type::hostile_spotted_near ) ) {
                 Creature *hostile_critter = is_hostile_very_close();
```

Comment:
It probably makes sense to move this added logic inside the `is_hostile_very_close` method, or even into the `game::is_hostile_within`. Perhaps, add a boolean parameter that enables this additional behavior. 

Two reasons for that:
1. Currently only one (random) creature is checked. If there are multiple creatures approaching, one of which is ranged, the logic may fail to warn the player
2. `game::is_hostile_very_close` is called from several other places, some of which may benefit from the new logic.

Reply:
Ok, done that in https://github.com/CleverRaven/Cataclysm-DDA/pull/45752/commits/a2c4e9febe9f39884d22f3ef38ccf1a9a7f8bdad.